src/Language/PureScript/Sugar/TypeClasses/Instances.hs

```diff
@@ -0,0 +1,48 @@
+-- |
+-- This module implements the desugaring pass which creates the compiler-generated
+-- names for type class instances that do not have one defined in the source code.
+--
+module Language.PureScript.Sugar.TypeClasses.Instances
+  ( desugarTypeClassInstanceNames
+  ) where
+
+import Prelude.Compat hiding (take)
+
+import           Control.Monad.Error.Class (MonadError(..))
+import           Control.Monad.Supply.Class
+import qualified Data.Map as M
+import           Data.Text (Text, take)
+import           Language.PureScript.Errors
+import           Language.PureScript.Names
+
+-- |
+-- Completes the name generation for type class instances that do not have
+-- a unique name defined in source code. All `Left Text` values
+-- will be converted to `Right Ident` values.
+--
+desugarTypeClassInstanceNames
+  :: (MonadSupply m, MonadError MultipleErrors m)
+  => Module
+  -> m (M.Map Text Ident, Module)
+desugarTypeClassInstanceNames (Module ss coms name decls exps) = do
+  results <- parU decls desugarInstName
+  let (desugared, instMaps) = foldr splitVals ([], []) results
+  pure (foldr M.union mempty instMaps, Module ss coms name desugared exps)
```

Comment:
Two comments here:
1. is there a better way to write `foldr splitVals ([], []) results`? I thought `sequence` would work, but I got a compiler error. So, I'm wondering if there's a different value I should be using or perhaps a type annotation
2. when `union`ing the `instMaps`, should an error be thrown here if the same key (i.e. same partially-generated instance name) is found multiple times? Right now, this will be caught as an internal error in the desugar pass, but here is where we could throw an error if this situation occurred.

Reply:
I'll give an actual review later, but isn't 1 just `unzip`?